### PR TITLE
[ci] Add sccache to ci building process on Linux and Mac jobs

### PR DESCRIPTION
**Pull request by lin-hitonami**
Related issue = #3528

### CODE REVIEW

File: .github/workflows/scripts/unix_build.sh [hunk 1]
```diff
@@ -4,7 +4,26 @@ python3 -m pip uninstall taichi taichi-nightly -y
 python3 -m pip install -r requirements_dev.txt
 python3 -m pip install -r requirements_test.txt
 git fetch origin master
-
+export SCCACHE_DIR=`pwd`/sccache_cache
+export SCCACHE_CACHE_SIZE="128M"
+echo sccache dir: $SCCACHE_DIR
+if [[ $OSTYPE == "linux-"* ]]; then
+    wget https://github.com/mozilla/sccache/releases/download/v0.2.15/sccache-v0.2.15-x86_64-unknown-linux-musl.tar.gz
+    tar -xzf sccache-v0.2.15-x86_64-unknown-linux-musl.tar.gz
```
Comment (frostming) [inline .github/workflows/scripts/unix_build.sh:12]: Is musl build usable on GNU linux?

Comment (bobcao3) [reply .github/workflows/scripts/unix_build.sh:12]: Seems like it can, circa https://www.musl-libc.org/faq.html

Comment (lin-hitonami) [reply .github/workflows/scripts/unix_build.sh:12]: It seems in this case it's working fine.

**Comment by lin-hitonami** [comment]
There seems to be some permission problem when I use `docker create -v sccache_cache:/home/dev/sccache_cache` to mount the cache to the docker. Even if I have already set the permissions all to 777 by `chmod -R 777 sccache_cache` outside the docker, the sccache still reports `Operation not permitted` error when reading the cache files. Copying the files to another directory, deleting the original files, and then moving the copied files to the original directory inside the docker can solve this problem. Generating a tarball outside the docker, copying the tarball into the docker by `docker cp`, and then extracting it inside the docker also solves the problem. I don't know if there is a better way to do this.

**Comment by frostming** [comment]
Why was windows change reverted, it is the biggest bottleneck 😢

**Comment by lin-hitonami** [comment]
Yes, but I have limited knowledge on windows and it seems that my configuration isn't working. It would be great if you could help us set up sccache on windows (maybe on another pr). @frostming

File: .github/workflows/scripts/unix_docker_build.sh [hunk 2]
```diff
@@ -7,12 +7,20 @@ PY=$1
 GPU_BUILD=$2
 PROJECT_NAME=$3
 CI_SETUP_CMAKE_ARGS=$4
-
+export SCCACHE_DIR=/home/dev/sccache_cache
```
Comment (ailzhang) [inline .github/workflows/scripts/unix_docker_build.sh:10]: Let's put sccache related setup code together? ;) Now it's split by conda & pip :P

File: .github/workflows/presubmit.yml [hunk 3]
```diff
@@ -184,14 +192,19 @@ jobs:
             exit 0
           fi
           mkdir -m777 wheel
+          mkdir -p sccache_cache
```
Comment (ailzhang) [inline .github/workflows/presubmit.yml:195]: As we discussed, let's add a note about why a tarball is needed here.

**Review by ailzhang — COMMENTED** [review]
LGTM with minor suggestions!
cc: @frostming

**Comment by frostming** [comment]
> There seems to be some permission problem when I use `docker create -v sccache_cache:/home/dev/sccache_cache` to mount the cache to the docker on the github-hosted runners

Can adding `:rw` to the mount solve the problem?

`docker create -v sccache_cache:/home/dev/sccache_cache:rw`

**Comment by lin-hitonami** [comment]
> > There seems to be some permission problem when I use `docker create -v sccache_cache:/home/dev/sccache_cache` to mount the cache to the docker on the github-hosted runners
> 
> Can adding `:rw` to the mount solve the problem?
> 
> `docker create -v sccache_cache:/home/dev/sccache_cache:rw`

It doesn't work. Even if I did not add rw, I can read, write, copy and delete the files. It seems that sccache performs some operations that only the file owner can do like maybe changing the mode of the files or something.

**Comment by frostming** [comment]
> > > There seems to be some permission problem when I use `docker create -v sccache_cache:/home/dev/sccache_cache` to mount the cache to the docker on the github-hosted runners
> > 
> > 
> > Can adding `:rw` to the mount solve the problem?
> > `docker create -v sccache_cache:/home/dev/sccache_cache:rw`
> 
> It doesn't work. Even if I do not add rw, I can read, write, copy and delete the files. It seems that sccache performs some operations that only the file owner can do like maybe changing the mode of the files or something.

Ok, let's use the `docker cp` solution

**Review by ailzhang — APPROVED** [review]
Thanks! This will help a looot! 
Btw I've been wondering why we have both unix_build.sh and unix_docker_build.sh and it make code a lot messier.... Gonna open an issue and see if anyone has context on that.

**Pull request merged**